spec.md

```diff
@@ -244,6 +244,58 @@ A common use of an externally defined module is to represent a Verilog module
 that will be written separately and provided together with FIRRTL-generated
 Verilog to downstream tools.
 
+## Functions
+
+A function has a name, list of arguments, and a list of statements representing
+a data-flow graph of stateless circuit components.  Each argument is specified 
+by its direction, which may be input or output, a name, and the data type of the 
+argument.
+
+Functions participate in width-inference (see [@sec:width-inference]) at 
+invocation sites in the same way modules to at instantiation sites.
+
+Functions cannot contain memories, registers, or instances.  This makes them
+"stateless"
+
+Functions can optional specify that they are "pure" functions.  These do not
+have side-effecting operations (such as printf).
+
+
+``` firrtl
+function MyPureFunction pure :
+  input foo : UInt
+  output bar : UInt
+  bar <= foo
+
+function MyFunction :
+  input clk: Clock
+  input halt: UInt<1>
+  stop(clk, halt, 42) : optional_name
+```
+
+## Intrinsics
+
+An intrinsic is a function. Intrinsics have all the components of a function
+except  a list of statements.  An intrinsic is a firrtl implementation provided 
+function.  The name of intrinsics are implementation defined.  A firrtl 
+implementation shall reject IR with unknown intrinsics.  Being functions,
+intrinsics are stateless.
+
+The widths of all intrinsic ports must be specified.  Width inference, described 
+in [@sec:width-inference], is not supported.
+
+Intrinsics can be "pure".  These do not have side-effects.  Whether a particular 
+intrinsic is pure or not is implementation defined and implementations shall 
+check that intrinsics declarations have the expected arguments, types, and 
+purity.
+
+``` firrtl
+intrinsic synopsys.4mux.u8 pure :
```

Comment:
If we allow vendor specific intrinsic, what happens if we do target other vendors? Do we want to introduce target-triple for FIRRTL and CIRCT land?

Reply:
What happens if you use an arm neon intrinsic when compiler C to amd64?

Reply:
We should probably also provide a default lowering to standard Comb/Seq ops for the intrinsics where that makes sense. For example, a dedicated mux4 intrinsic should provide a lowering to a `comb.mux` for all those tools and passes that don't need the added detail of the intrinsic. But agreed, some will be vendor/target-specific and require the frontend to just not emit them.